internal/rbd/replicationcontrollerserver.go

```diff
@@ -0,0 +1,357 @@
+// volumeID,If the image is present it will enable the mirroring based on the
+// user provided information.
+// TODO: create new Replication controller struct for the replication operations.
+func (cs *ControllerServer) EnableVolumeReplication(ctx context.Context,
+	req *replication.EnableVolumeReplicationRequest,
+) (*replication.EnableVolumeReplicationResponse, error) {
+	rbdVol, cr, err := cs.getVolumeFromID(ctx, req.GetVolumeId(), req.GetSecrets())
+	defer cs.cleanup(rbdVol, cr)
+	if err != nil {
+		return nil, err
+	}
+	// extract the mirroring mode
+	mirroringMode, err := getMirroringMode(ctx, req.GetParameters())
+	if err != nil {
+		return nil, err
+	}
+
+	mirroringInfo, err := rbdVol.getImageMirroringInfo(ctx)
+	if err != nil {
+		util.ErrorLog(ctx, "failed to get mirroring info of %s with error %s", rbdVol.String(), err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if mirroringInfo.State != librbd.MirrorImageEnabled {
+		err = rbdVol.enableImageMirroring(ctx, mirroringMode)
+		if err != nil {
+			util.ErrorLog(ctx, "failed to enable mirroring of %s with error %s", rbdVol.String(), err.Error())
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+	}
+	return &replication.EnableVolumeReplicationResponse{}, nil
+}
+
+// DisableVolumeReplication extracts the RBD volume information from the
+// volumeID, If the image is present and the mirroring is enabled on the RBD
+// image it will disable the mirroring.
+func (cs *ControllerServer) DisableVolumeReplication(ctx context.Context,
+	req *replication.DisableVolumeReplicationRequest,
+) (*replication.DisableVolumeReplicationResponse, error) {
+	rbdVol, cr, err := cs.getVolumeFromID(ctx, req.GetVolumeId(), req.GetSecrets())
+	defer cs.cleanup(rbdVol, cr)
+	if err != nil {
+		return nil, err
+	}
+
+	// extract the force option
+	force, err := getForceOption(ctx, req.GetParameters())
+	if err != nil {
+		return nil, err
+	}
+
+	mirroringInfo, err := rbdVol.getImageMirroringInfo(ctx)
+	if err != nil {
+		util.ErrorLog(ctx, "failed to get mirroring info of %s with error %s", rbdVol.String(), err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	switch mirroringInfo.State {
+	case librbd.MirrorImageEnabled:
+		if !force && !mirroringInfo.Primary {
+			return nil, status.Error(codes.InvalidArgument, "image is in non-primary state")
+		}
+		err = rbdVol.disableImageMirroring(ctx, force)
+		if err != nil {
+			util.ErrorLog(ctx, "failed to disable mirroring of %s with error %s", err.Error())
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+		// TODO: check do we need to return abort after disabling the mirroring
+		return &replication.DisableVolumeReplicationResponse{}, nil
+	case librbd.MirrorImageDisabled:
+		// image is already in disabled state
+	default:
+		// TODO: use string instead of int for returning valid error message
+		return nil, status.Errorf(codes.InvalidArgument, "image is in %d Mode", mirroringInfo.State)
```

Comment:
@phlogistonjohn should go-ceph offer a way to translate constants like `MirrorImageDisabled` to a string so that consumers can better log it?